keps/sig-api-machinery/20200309-consistent-resource-versions-semantics.md

```diff
@@ -0,0 +1,216 @@
+  - [Alternative: Use syntax in the query string](#alternative-use-syntax-in-the-query-string)
+<!-- /toc -->
+
+## Summary
+
+Make resource version semantics consistent for list and get requests regardless of
+pagination.
+
+## Motivation
+
+Resource version semantics are inconsistent when using pagination. When a list
+request is made with a resourceVersion but no limit, “Not older than” semantics
+apply, but once a limit is set, “Exact” semantics apply. See [API Concepts: Resource Version](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions)
+for details about the semantics.
+
+The inconsistency is confusing and problematic. A client expecting “Not older
+than” semantics has no reason to expect the “410 Gone” HTTP responses from the
+server that can be sent when “Exact” semantics are applied and when the
+requested resource version is older than what is available in watch cache or the
+etcd compaction window. Even if the author of code to make a list request was
+aware of this inconsistency, the code to make a list request might be separated
+from the code setting a limit on the request. For example, ListerWatchers can be
+layered in to add pagination limits.
+
+This was discussed in [Feb 12th, 2020 SIG API machinery bi-weekly meeting](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#bookmark=id.3kvpricxohe8).
+
+### Goals
+
+- Improve API to make resource version semantics consistent.
+- Backward compaibility. In particular, support the case of a client making a request using the
+  updated API against an server serving the old API.
+
+## Proposal
+
+### Add a ResourceVersionMatch query parameter
+
+Add an optional `ResourceVersionMatch` paramater to `ListOptions` and
+`GetOptions` with the enumeration values:
+
+```
+// ResourceVersionMatch specifies how the ResourceVersion parameter is applied. ResourceVersionMatch
+// may only be set if ResourceVersion is also set.
+type ResourceVersionMatch string
+
+const (
+	// ResourceVersionMatchNotOlderThan matches data at least as new as the provided
+	// ResourceVersion. The newest available data is preferred, but any data not
+	// older than this ResourceVersion may be served.
+	// This guarantees that ResourceVersion in the ListMeta is not older than the requested
+	// ResourceVersion, but does not make any guarantee about the ResourceVersion in the ObjectMeta
+	// of the list items since ObjectMeta.ResourceVersion tracks when an object was last updated,
+	// not how up-to-date the object is when served.
+	ResourceVersionMatchNotOlderThan ResourceVersionMatch = "NotOlderThan"
+	// ResourceVersionMatchExact matches data at the exact ResourceVersion
+	// provided. If the provided ResourceVersion is unavailable, the server responds with
+	// HTTP 410 “Gone”.
+	// This guarantees that ResourceVersion in the ListMeta is the same as the requested
```

Comment:
same here - focus just on LIST request, but this will be used in other requests too